### PR TITLE
Add Stochastic Weight Averaging Optimizer.

### LINKED ISSUE #592: Add Stochastic Weight Averaging (SWA) Optimizer
__Describe the feature and the current behavior/state.__

SWA has been shown to be a simple method to improve model performance when used as a wrapper around other optimizers and averaging results from different points of trajectory of the inner optimizer.

__Relevant information__

- Are you willing to contribute it (yes/no): yes, I have an implementation that I should be able to port to the repository fairly easily. The implementation looks pretty similar to the implementation for the Lookahead Optimizer that was recently added to addons.
- Are you willing to maintain it going forward? (yes/no): yes
- Is there a relevant academic paper? (if so, where):  yes, [Averaging Weights Leads to Wider Optima and Better Generalization](https://arxiv.org/abs/1803.05407)
- Is there already an implementation in another framework? (if so, where): yes, [Pytorch](https://github.com/pytorch/contrib/tree/master/torchcontrib/optim) and [fast.ai](https://github.com/fastai/fastai/blob/master/old/fastai/swa.py) (this is in old right now because they seem to be prepping for a merge to the new version).
- Was it part of tf.contrib? (if so, where): No 

__Which API type would this fall under (layer, metric, optimizer, etc.)?__

This would fall under the optimizer API.

__Who will benefit with this feature?__

SWA works like any other optimizer so it can be a good addition to standard Adam, SGD or RMSProp to help squeeze out a little bit of extra performance.

### PR DESCRIPTION
**Pull request by shreyashpatodia**
Adding Stochastic Weight Averaging Optimizer.

Notes:

* The optimizer needs the calling of the `assign_average_vars()` at the end of training to assign the averaged values to the model variables. Docstring for `assign_average_vars()` contains an example on how to do this.

* Averaging starts happening at step `start_averaging` and then happens at every `average_period` steps.

Closes #592

### CODE REVIEW

**Comment by shreyashpatodia** [comment]
@facaiy @WindQAQ @Squadrick - I've run the code-format, sanity-check and cpu test locally and everything looks good. Is there anything else I should do for this?

Thanks!

**Comment by shreyashpatodia** [comment]
@seanpmorgan could you run the tests again.

One of my tests was failing due to an error of 1.1e-4 instead of 1e-4 (where 1e-4 was the threshold). I've slightly increased the threshold and added more epochs of training to hopefully remove this from being a flaky test.

File: tensorflow_addons/optimizers/stochastic_weight_averaging.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+    def get_config(self):
+        config = {
+            'optimizer': tf.keras.optimizers.serialize(self._optimizer),
+            'average_period': self._serialize_hyperparameter('average_period'),
+            'start_averaging':
+            self._serialize_hyperparameter('start_averaging')
+        }
+        base_config = super(SWA, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
```
Comment (yxliang01) [inline tensorflow_addons/optimizers/stochastic_weight_averaging.py]:
```suggestion
        return { **base_config, **config }
```

The suggestion is more pythonic(use of dict comprehension) and potentially faster? Since there would be less conversion.

**Comment by yxliang01** [comment]
@googlebot I fixed it.

**Comment by shreyashpatodia** [comment]
@seanpmorgan looks like the change suggested by @yxliang01 isn't compatible with Python2, reverted back to previous

**Comment by shreyashpatodia** [comment]
@seanpmorgan might need to run the tests again because I accidently closed the PR.

Sorry for the inconvenience!

**Comment by shreyashpatodia** [comment]
@facaiy @Squadrick @WindQAQ would it be possible for you guys to take a look at this PR?

**Comment by shreyashpatodia** [comment]
@facaiy @WindQAQ @Squadrick would it be possible for you guys to take a look at this PR?

File: tensorflow_addons/optimizers/README.md [hunk 2]
```diff
@@ -10,6 +10,7 @@
 | moving_average | Dheeraj R. Reddy | dheeraj98reddy@gmail.com |
 | rectified_adam | Zhao Hanguang | cyberzhg@gmail.com |
 | weight_decay_optimizers |  Phil Jund | ijund.phil@googlemail.com   |
+| stochastic_weight_averaging | Shreyash Patodia | patodiashreyash32@gmail.com |
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/README.md]: Alphabetical order

File: tensorflow_addons/optimizers/README.md [hunk 3]
```diff
@@ -22,6 +23,7 @@
 | moving_average | MovingAverage | |
 | rectified_adam | RectifiedAdam | https://arxiv.org/pdf/1908.03265v1.pdf |
 | weight_decay_optimizers | SGDW, AdamW, extend_with_decoupled_weight_decay | https://arxiv.org/pdf/1711.05101.pdf |
+| stochastic_weight_averaging | SWA | https://arxiv.org/abs/1803.05407.pdf |
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/README.md]: ditto

File: tensorflow_addons/optimizers/stochastic_weight_averaging.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import tensorflow as tf
+from tensorflow_addons.utils import keras_utils
+
+
+@keras_utils.register_keras_custom_object
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/stochastic_weight_averaging.py]: Replace with `@tf.keras.utils.register_keras_serializable(package='Addons')`. See 2b070a15fb516e6b3c1cca3898ae611024d7481d.

File: tensorflow_addons/optimizers/stochastic_weight_averaging_test.py [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+            update = optimizer.assign_average_vars([var_0, var_1])
+            self.evaluate(update)
+        else:
+            optimizer.assign_average_vars([var_0, var_1])
+
+        self.assertAllClose(var_0.read_value(), [0.8, 0.8])
+        self.assertAllClose(var_1.read_value(), [1.8, 1.8])
+
+    def test_fit_simple_linear_model(self):
+        np.random.seed(0x2019)
```
Comment (Squadrick) [inline tensorflow_addons/optimizers/stochastic_weight_averaging_test.py]: Add a seed for TensorFlow as well.

**Comment by shreyashpatodia** [comment]
@googlebot I signed it!

File: tensorflow_addons/optimizers/stochastic_weight_averaging_test.py (hunk 5 above)
Comment (shreyashpatodia) [reply tensorflow_addons/optimizers/stochastic_weight_averaging_test.py]: Done.

**Comment by shreyashpatodia** [comment]
@Squadrick it might be worth running the tests again. My time at Google has come to an end so I've made the changes via the github editor for now. If we could run the tests online and everything is fine otherwise I'll setup the project on my laptop.

Thanks!

**Comment by shreyashpatodia** [comment]
@seanpmorgan could you run the tests again?

Thanks!

**Comment by shreyashpatodia** [comment]
@seanpmorgan - sorry about having to make you run the test multiple times. I was using the github editor and it turns out it was adding whitespaces in empty lines. I finally bit the bullet and built the project on my computer to be able to see why the sanity check was actually failing (the failure reason in the logs isn't particularly useful.

**Comment by seanpmorgan** [comment]
> @seanpmorgan - sorry about having to make you run the test multiple times. I was using the github editor and it turns out it was adding whitespaces in empty lines. I finally bit the bullet and built the project on my computer to be able to see why the sanity check was actually failing (the failure reason in the logs isn't particularly useful.

No worries....  it should really run automatically anyway!

**Comment by shreyashpatodia** [comment]
> > @seanpmorgan - sorry about having to make you run the test multiple times. I was using the github editor and it turns out it was adding whitespaces in empty lines. I finally bit the bullet and built the project on my computer to be able to see why the sanity check was actually failing (the failure reason in the logs isn't particularly useful.
> 
> No worries.... it should really run automatically anyway!

I'd be happy to setup automatic testing (unless there is a good reason to not automatically test things).

**Comment by shreyashpatodia** [comment]
@Squadrick everything should be done.

Thanks for your review!
